### DISCUSSION TITLE
PersistJobDataAfterExecution not persisting data

### DISCUSSION THREAD

**Discussion by CrazyMoka** [quartznet/quartznet#1685] [Q&A]
I'm trying to keep track of execution count and other data but I can't seem to get it to work.

                _job = JobBuilder.Create(jobClass)
                    .WithIdentity(jobKey)
                    .UsingJobData("ExecutionCount",1)
                    .StoreDurably(true).Build();

In the Job class

				JobKey key = context.JobDetail.Key;

				JobDataMap dataMap = context.MergedJobDataMap;

				string message = dataMap.GetString("Message") ?? "";

				int user_id = dataMap.GetIntValueFromString("User_Id");
				
				if (dataMap.ContainsKey("ExecutionCount"))
				{
					this.runCount = dataMap.GetInt("ExecutionCount");
				}

				passErrorsMessages(context);
				this.runCount++;
				dataMap.Put("ExecutionCount", runCount);

logger.LogInformation("Run Count " + runCount); is always 1 and never increments.

What am I doing wrong?

Thank you

**Comment by lahma** [comment]
Maybe enabling logging could help here. There's also example in Quartz.Examples showing it working.

**Reply by CrazyMoka** [reply]
Hi @lahma 
I have logging on. And I copied the example from Quartz.Examples. The only thing I didn't do was make a private varible for the string Key.
Everytime it runs it incremetns teh counter and it shows that it in the datamap but when it runs it again and gets it out of the datamap its always set back to 1. So no persistence,

I have a schedule listener and no errors are being thrown.

I'll try coping it exactly I guess see what happens.

**Reply by lahma** [reply]
And your job has `PersistJobDataAfterExecutionAttribute`?

**Reply by CrazyMoka** [reply]
I have this:
	[PersistJobDataAfterExecution]
	[DisallowConcurrentExecution]
	public class TestJob : IJob
	
	So yes.

**Reply by thedangler** [reply]
It turns out if you use context.MergedJobDataMap;
The data is not preserved.  Is that by design?

**Reply by lahma** [reply]
Ah sorry, I totally missed that you were using `MergedJobDataMap`. From documentation:

https://github.com/quartznet/quartznet/blob/6c81befdb55bda55b6dbc596630c2a19c1aef973/src/Quartz/IJobExecutionContext.cs#L67-L68

Quartz wouldn't know whether trigger data map or job data map should be updated as it's *merged* from both of them.

**Reply by thedangler** [reply]
Ok, How would I maintain data in triggers?
As I have Jobs that run multiple triggers so I pass variables to the trigger not the job.  But I need to update error info in the trigger data but I can't seem to find a way to do it. 

Can't seem to update the JPB_DATA field in the Quartz trigger table

**Reply by lahma** [reply]
Hmm.. I have never encountered such use case before that one would like to update trigger data map. It's generally considered as static just like the trigger scheduling configuration. So to update either you would need to reschedule the job with IScheduler API. I don't know what will happen if you mutate it while running the job for the specific trigger though...

Maybe you would need separate trigger history/status table that you can use to track such information.

**Comment by thedangler** [comment]
I'm using db to store data.
Persisting simply does not work even when trying their example.

The variables to indicate if it should save are all true, like is Dirty.  But yet it doesn't do anything.

Any other way I can debut this?
